diff --git a/private/rules/jvm_import.bzl b/private/rules/jvm_import.bzl
index 0fbbd10..7214816 100644
--- a/private/rules/jvm_import.bzl
+++ b/private/rules/jvm_import.bzl
@@ -26,6 +26,8 @@ def _jvm_import_impl(ctx):
     if not ctx.attr.jar:
         print(ctx.attr.name, "The `jars` attribute is deprecated and will be removed in `rules_jvm_external` 7.0. Please use the `jar` attribute instead.")

+    java_runtime = ctx.toolchains["@bazel_tools//tools/jdk:runtime_toolchain_type"].java_runtime
+    add_jar_manifest_entry = ctx.file._add_jar_manifest_entry
     injar = ctx.file.jar if ctx.attr.jar else ctx.files.jars[0]

     if ctx.attr._stamp_manifest[StampManifestProvider].stamp_enabled:
@@ -33,11 +35,15 @@ def _jvm_import_impl(ctx):
         args = ctx.actions.args()
         args.add_all(["--source", injar, "--output", outjar])
         args.add("--manifest-entry", ctx.label, format = "Target-Label:%s")
+        java_args = ctx.actions.args()
+        java_args.add("-jar")
+        java_args.add(add_jar_manifest_entry)
         ctx.actions.run(
-            executable = ctx.executable._add_jar_manifest_entry,
-            arguments = [args],
-            inputs = [injar],
+            executable = java_runtime.java_executable_exec_path,
+            arguments = [java_args, args],
+            inputs = [injar, add_jar_manifest_entry],
             outputs = [outjar],
+            tools = java_runtime.files,
             mnemonic = "StampJarManifest",
             progress_message = "Stamping the manifest of %{label}",
         )
@@ -59,13 +65,17 @@ def _jvm_import_impl(ctx):
     # Make sure the compile jar is safe to compile with
     args.add("--make-safe")

+    java_args = ctx.actions.args()
+    java_args.add("-jar")
+    java_args.add(add_jar_manifest_entry)
     ctx.actions.run(
-        executable = ctx.executable._add_jar_manifest_entry,
-        arguments = [args],
-        inputs = [outjar],
+        executable = java_runtime.java_executable_exec_path,
+        arguments = [java_args, args],
+        inputs = [outjar, add_jar_manifest_entry],
         outputs = [compilejar],
+        tools = java_runtime.files,
         mnemonic = "CreateCompileJar",
-        progress_message = "Creating compile jar for %s" % ctx.label,
+        progress_message = "Creating compile jar for %{label}",
     )

     additional_providers = []
@@ -129,9 +139,9 @@ jvm_import = rule(
             doc = "URL from where `jar` will be downloaded from.",
         ),
         "_add_jar_manifest_entry": attr.label(
-            executable = True,
+            allow_single_file = True,
             cfg = "exec",
-            default = "//private/tools/java/com/github/bazelbuild/rules_jvm_external/jar:AddJarManifestEntry",
+            default = "//private/tools/java/com/github/bazelbuild/rules_jvm_external/jar:AddJarManifestEntry_deploy.jar",
         ),
         "_stamp_manifest": attr.label(
             default = "@rules_jvm_external//settings:stamp_manifest",
@@ -139,4 +149,5 @@ jvm_import = rule(
     },
     implementation = _jvm_import_impl,
     provides = [JavaInfo],
+    toolchains = ["@bazel_tools//tools/jdk:runtime_toolchain_type"],
 )
